### PR TITLE
feat(language): VSCODE-24: Setup language Server

### PR DESCRIPTION
**Pull request by imlucas**
## Description

This PR provides the initial setup for MongoDB Language Server; primarily, it is documentation, dependencies and test fixture setup we'll add user-facing features on top of in the future. I also added a static mongodb language configuration in `./language` for the editor.

A test fixture workspace has been added under `src/test/fixture`.

### Reviewer Notes

- See the [./src/language/README.md](https://github.com/mongodb-js/vscode/blob/568a653941dbd4d8a071b61881bbba898e7f3071/src/language/README.md). Is anything unclear or missing detail?
- 

### Add mongodb language configuration

Just the JS static language config for now so `.mongodb` gets some basic JS bracket matching/folding/etc.

https://code.visualstudio.com/api/language-extensions/language-configuration-guide

> The [`contributes.languages`](/api/references/contribution-points#contributes.languages) Contribution Point allows you to define a language configuration that controls the following Declarative Language Features:
>
> - Comment toggling
> - Brackets definition
> - Autoclosing
> - Autosurrounding
> - Folding
> - Word pattern
> - Indentation Rules

## Todo

- [ ] Add stubs interfaces for document → connection API's

### Checklist
- [x] New tests and/or benchmarks are included
- [x] Documentation is changed or added

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [x] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by imlucas** [comment]
Screenshot 2020-03-11 12 04 42
Screenshot 2020-03-11 12 41 23

**Comment by Anemy** [comment]
Hmmm I got this same mac error on my pr branch:
Screen Shot 2020-03-12 at 9 30 14 AM

**Comment by imlucas** [comment]
@Anemy yeah same on data-service

**Comment by imlucas** [comment]
@Anemy here's the data-service ticket for async stack corrupted https://jira.mongodb.org/browse/COMPASS-4087

**Comment by imlucas** [comment]
@alenakhineika @Anemy could you have a look at this and let me know what you think? Most details are in https://github.com/mongodb-js/vscode/blob/568a653941dbd4d8a071b61881bbba898e7f3071/src/language/README.md

The actual `.ts` is boilerplate from samples. If this covers it, we should make tickets for actual impls behind completion etc.

File: src/language/index.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import { workspace, ExtensionContext } from 'vscode';
+
+import {
+  LanguageClient,
+  LanguageClientOptions,
+  ServerOptions,
+  TransportKind
+} from 'vscode-languageclient';
+
+let client: LanguageClient;
```
Comment (Anemy) [inline src/language/index.ts]: Can this live on the `LanguageServerController` class?

File: src/language/server.ts [hunk 2]
```diff
@@ -0,0 +1,349 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
```
Comment (Anemy) [inline src/language/server.ts]: nit - I think we want to remove these?

File: src/language/index.ts (hunk 1 above)
Comment (imlucas) [reply src/language/index.ts]: Done.

File: src/language/server.ts (hunk 2 above)
Comment (imlucas) [reply src/language/server.ts]: Done.

**Comment by imlucas** [comment]
>  really like the potential integration with the schema - I see that being saving a lot of headaches with developers. 

Yes. This will enable caching along with not blocking UI.

> Do you feel like you're in a place to create out the tickets for the language server - or a couple starting ones?
Yes.

> Clarifying for my own head, this work is setting up the Language Controller, which currently we don't activate yet, but that will be starting src/language/server.ts on activation which handles the actual language feature behaviors, like code completions (hopefully schema/collection/database name completions) and prettier/ more general language features. The server will be listening to the multiple document events which we can then fill out the behaviors with.

Yes. See https://github.com/zhaoyi0113/mongodb-runner/tree/master/server/src for examples.

**Comment by imlucas** [comment]
@Anemy since rebasing against master this morning I can no longer compile and I get the 54 errors below. Any ideas for unblocking?

```
src/connectionController.ts:36:39 - error TS1005: ',' expected.

36   connectWithConnectionId(connectionId:<any>): void {
                                         ~

src/connectionController.ts:36:45 - error TS1109: Expression expected.

36   connectWithConnectionId(connectionId:<any>): void {
                                               ~

src/connectionController.ts:36:46 - error TS1005: ';' expected.

36   connectWithConnectionId(connectionId:<any>): void {
                                                ~

src/connectionController.ts:37:11 - error TS1005: ':' expected.

37     throw new Error("Method not implemented.");
             ~~~

src/connectionController.ts:37:47 - error TS1005: ',' expected.

37     throw new Error("Method not implemented.");
                                                 ~

src/connectionController.ts:39:32 - error TS1005: ',' expected.

39   renameConnection(connectionId:<any>): void {
                                  ~

src/connectionController.ts:39:38 - error TS1109: Expression expected.

39   renameConnection(connectionId:<any>): void {
                                        ~

src/connectionController.ts:39:39 - error TS1005: ';' expected.

39   renameConnection(connectionId:<any>): void {
                                         ~

src/connectionController.ts:40:11 - error TS1005: ':' expected.

40     throw new Error("Method not implemented.");
             ~~~

src/connectionController.ts:40:47 - error TS1005: ',' expected.

40     throw new Error("Method not implemented.");
                                                 ~

src/connectionController.ts:42:27 - error TS1005: ';' expected.

42   getActiveConnectionId() {
                             ~

src/connectionController.ts:45:34 - error TS1005: ';' expected.

45   getActiveConnectionDriverUrl() {
                                    ~

src/editors/editorsController.ts:22:27 - error TS1005: ',' expected.

22   onViewDocument(namespace: any, documentId: any): Promise<boolean> {
                             ~

src/editors/editorsController.ts:22:44 - error TS1005: ',' expected.

22   onViewDocument(namespace: any, documentId: any): Promise<boolean> {
                                              ~

src/editors/editorsController.ts:22:50 - error TS1005: ';' expected.

22   onViewDocument(namespace: any, documentId: any): Promise<boolean> {
                                                    ~

src/editors/editorsController.ts:23:11 - error TS1005: ':' expected.

23     throw new Error("Method not implemented.");
             ~~~

src/editors/editorsController.ts:23:47 - error TS1005: ',' expected.

23     throw new Error("Method not implemented.");
                                                 ~

src/explorer/connectionTreeItem.ts:17:3 - error TS1005: ',' expected.

17   connectionId(connectionId: any): Promise<boolean> {
     ~~~~~~~~~~~~

src/explorer/connectionTreeItem.ts:17:28 - error TS1005: ',' expected.

17   connectionId(connectionId: any): Promise<boolean> {
                              ~

src/explorer/connectionTreeItem.ts:17:34 - error TS1005: ',' expected.

17   connectionId(connectionId: any): Promise<boolean> {
                                    ~

src/explorer/connectionTreeItem.ts:18:5 - error TS1068: Unexpected token. A constructor, method, accessor, or property was expected.

18     throw new Error("Method not implemented.");
       ~~~~~

src/explorer/connectionTreeItem.ts:19:3 - error TS1128: Declaration or statement expected.

19   }
     ~

src/explorer/connectionTreeItem.ts:20:14 - error TS1005: ';' expected.

20   implements TreeItemParent, vscode.TreeDataProvider<ConnectionTreeItem> {
                ~~~~~~~~~~~~~~

src/explorer/connectionTreeItem.ts:21:58 - error TS1005: ',' expected.

21   contextValue = ConnectionItemContextValues.disconnected;
                                                            ~

src/explorer/connectionTreeItem.ts:23:33 - error TS1005: ']' expected.

23   private _childrenCache: { [key: string]: DatabaseTreeItem };
                                   ~

src/explorer/connectionTreeItem.ts:23:41 - error TS1005: ',' expected.

23   private _childrenCache: { [key: string]: DatabaseTreeItem };
                                           ~

src/explorer/connectionTreeItem.ts:23:42 - error TS1136: Property assignment expected.

23   private _childrenCache: { [key: string]: DatabaseTreeItem };
                                            ~

src/explorer/connectionTreeItem.ts:23:62 - error TS1005: ',' expected.

23   private _childrenCache: { [key: string]: DatabaseTreeItem };
                                                                ~

src/explorer/connectionTreeItem.ts:24:35 - error TS1005: ',' expected.

24   _childrenCacheIsUpToDate = false;
                                     ~

src/explorer/connectionTreeItem.ts:26:54 - error TS1005: ',' expected.

26   private _connectionController: ConnectionController;
                                                        ~

src/explorer/connectionTreeItem.ts:27:23 - error TS1005: ',' expected.

27   connectionId: string;
                         ~

src/explorer/connectionTreeItem.ts:29:22 - error TS1005: ',' expected.

29   isExpanded: boolean;
                        ~

src/explorer/connectionTreeItem.ts:61:3 - error TS1005: ',' expected.

61   get tooltip(): string {
     ~~~

src/explorer/connectionTreeItem.ts:65:3 - error TS1005: ',' expected.

65   get description(): string {
     ~~~

src/explorer/connectionTreeItem.ts:87:3 - error TS1005: ',' expected.

87   getTreeItem(element: ConnectionTreeItem): ConnectionTreeItem {
     ~~~~~~~~~~~

src/explorer/connectionTreeItem.ts:91:3 - error TS1005: ',' expected.

91   getChildren(): Thenable<any[]> {
     ~~~~~~~~~~~

src/explorer/connectionTreeItem.ts:148:3 - error TS1005: ',' expected.

148   get iconPath():
      ~~~

src/explorer/connectionTreeItem.ts:169:3 - error TS1005: ',' expected.

169   onDidCollapse(): void {
      ~~~~~~~~~~~~~

src/explorer/connectionTreeItem.ts:174:3 - error TS1005: ',' expected.

174   onDidExpand(): Promise<boolean> {
      ~~~~~~~~~~~

src/explorer/connectionTreeItem.ts:199:3 - error TS1005: ',' expected.

199   resetCache(): void {
      ~~~~~~~~~~

src/explorer/connectionTreeItem.ts:204:3 - error TS1005: ',' expected.

204   getChildrenCache(): { [key: string]: DatabaseTreeItem } {
      ~~~~~~~~~~~~~~~~

src/explorer/connectionTreeItem.ts:208:3 - error TS1005: ',' expected.

208   async onAddDatabaseClicked(
      ~~~~~

src/explorer/documentTreeItem.ts:5:22 - error TS1005: ',' expected.

5   namespace(namespace: <any>, documentId:<any>): Promise<boolean> {
                       ~

src/explorer/documentTreeItem.ts:5:29 - error TS1109: Expression expected.

5   namespace(namespace: <any>, documentId:<any>): Promise<boolean> {
                              ~

src/explorer/documentTreeItem.ts:5:41 - error TS1005: ',' expected.

5   namespace(namespace: <any>, documentId:<any>): Promise<boolean> {
                                          ~

src/explorer/documentTreeItem.ts:5:47 - error TS1109: Expression expected.

5   namespace(namespace: <any>, documentId:<any>): Promise<boolean> {
                                                ~

src/explorer/documentTreeItem.ts:5:48 - error TS1005: ';' expected.

5   namespace(namespace: <any>, documentId:<any>): Promise<boolean> {
                                                 ~

src/explorer/documentTreeItem.ts:6:11 - error TS1005: ':' expected.

6     throw new Error("Method not implemented.");
            ~~~

src/explorer/documentTreeItem.ts:6:47 - error TS1005: ',' expected.

6     throw new Error("Method not implemented.");
                                                ~

src/explorer/documentTreeItem.ts:8:23 - error TS1005: ',' expected.

8   documentId(namespace: any, documentId: any): Promise<boolean> {
                        ~

src/explorer/documentTreeItem.ts:8:40 - error TS1005: ',' expected.

8   documentId(namespace: any, documentId: any): Promise<boolean> {
                                         ~

src/explorer/documentTreeItem.ts:8:46 - error TS1005: ';' expected.

8   documentId(namespace: any, documentId: any): Promise<boolean> {
                                               ~

src/explorer/documentTreeItem.ts:9:11 - error TS1005: ':' expected.

9     throw new Error("Method not implemented.");
            ~~~

src/explorer/documentTreeItem.ts:9:47 - error TS1005: ',' expected.

9     throw new Error("Method not implemented.");
                                                ~
```

**Comment by Anemy** [comment]
Hmmm those looks like merge conflicts that might have resolved incorrectly. Testing on master is smooth.

**Comment by imlucas** [comment]
@Anemy thanks for the catch. I've split this out to #46 and #47. We'll whiteboard on Monday rather than me just making tickets :)

**Pull request closed without merging**